atools/_memoize_decorator.py

```diff
@@ -190,7 +193,29 @@ def finalize_memo(self, memo: _Memo, key: Union[int, str]) -> Any:
                 )
             return memo.memo_return_state.value
 
-    def get_hashed_key(self, key: Tuple[Hashable]) -> Union[int, str]:
+    def get_hashed_key(self, key: Union[Hashable, Tuple[Hashable]]) -> Union[int, str]:
+        # Make a best-effort attempt to warn developers if they use a keygen
+        # function that relies on the default hash implementation for object.
+        if not self._has_checked_keygen:
+            if isinstance(key, tuple):
+                for component in key:
+                    # The default  __hash__ implementation can re-use hash values
+                    # if the original object is deleted.
+                    if hasattr(type(component), "__hash__") and type(component).__hash__ == object.__hash__:
+                        warnings.warn(
+                            "`{}` is memoized with {} objects in the cache key. "
+                            "However, {} uses the default hash function for "
+                            "objects, which can produce duplicate cache lookups "
+                            "in any given Python process. Consider adding a "
+                            "custom __hash__ method to that class".format(
+                                self.fn.__name__,
+                                type(component).__name__,
+                                type(component).__name__,
+                            )
+                        )
+
+            object.__setattr__(self, '_has_checked_keygen', True)
```

Comment:
Apologies for this kludge :-) The goal is to only perform the check once per decorated function (for efficiency's sake), but I couldn't find a better way to store this state.